### ISSUE TITLE
How to edit 3D meshes?

### ISSUE THREAD

**Issue by brylie** [jagenjo/webglstudio.js#39]
We are researching 3D editors for WebGL that allow users to create and edit 3D meshes. E.g. we would like to be able to do things like move edges/verticies, divide faces, etc.

After adding a default cube, how do I edit its mesh?

**Comment by jagenjo** [comment]
WebGLStudio doesnt have mesh editing capabilities (maybe one day...)

Its main goal is to create 3D applications, but meshes must be imported from any 3D software.
I suggest you to check clara.io which is more a 3D meshes editor.

Cheers

**Comment by brylie** [comment]
I had found clara.io, but it is not open source. In effect, this is a feature request to be able to edit meshes with WebGL Studio, but I wasn't sure to what extent the ability existed. Please re-open the issue, since it is a feature request.

**Comment by jagenjo** [comment]
ok, but  Im the sole developer of WebGLStudio and I dont have plans to add that feature (it will take one year of work to develop it), so do not expect it to be available in quite some time.

**Comment by brylie** [comment]
That makes sense. We may be able to help. What would be some sub-tasks for the feature?

**Comment by jagenjo** [comment]
I can make an skeleton for the mesh editor, but you will have to code the actions like:
- select one or multiple vertex, edge or face
- move, rotate, scale
- subdivision, extrusion, etc

The main problem is that my mesh structure is not meant for editing, in the sense that it just packs all the vertices in an array (focused on rendering performance), and to edit meshes you usually want to have a structure that has info like neighbour triangles, etc.

If you do not have experience it could take some time, thats why I never bothered trying, but if you are commited to the idea Im willing to help.

**Comment by brylie** [comment]
> The main problem is that my mesh structure is not meant for editing, in the sense that it just packs all the vertices in an array (focused on rendering performance), and to edit meshes you usually want to have a structure that has info like neighbour triangles, etc.

Might it be good to add an 'edit mode' toggle, so the default representation could be performance optimized, while edit mode would switch to a more topology-based representation? E.g. this is similar to the Blender modelling experience.

**Comment by jagenjo** [comment]
yes, thats what I was thinking, it shouldnt be hard to have a property inside every mesh for a geometric representation of itself (triangles, edges, vertices), and methods to convert from one to the other.

I was just saying it to showcase the complexities of editing.

**Comment by jagenjo** [comment]
because that representation is totally tied to the editor itself, the person in charge of doing it should decide how he wants to structure it. I do not have experience with mesh editing so I dont know what are the best way to structure it.

The main problem is that for complex meshes (with thousands of vertices) if that structure is not accelerated (using some sort of spatial partitioning) it could be a little bit slow because for every click you have to traverse all the faces. Although we could asume small meshes to start.

**Comment by coderofsalvation** [comment]
[blender3d](https://blender.org) is the king of mesh editing.
Unity3D neither supports such features, hence they just point people to blender3d.
I would think focusing on mesh-editing would be a waste of precious resources.
Blender3d & webglstudio is a perfect👌 combination, both free and opensource.

**Comment by kalwalt** [comment]
I agree with you @coderofsalvation, of someone want to edit a mesh modify it with blender3d.
